# Notebook 28 — GRU Diagnostic Analysis

**Date**: 2026-03-01

**Contexte**: GRU warm+no-capping = 49.5% Hit@1 (tool=55%, cap=43%). Champion 57.6% était gonflé par MPC=30.

**Mystère résolu**: `benchmark-e2e.ts` (L1120-1137) persiste les 920 tools SHGAT-enrichis en DB
via `tool_embedding.shgat_embedding`. MAIS l'enrichissement réel dépend de la connectivité SHGAT :
- **Hierarchical** (195 tools dans caps) : L2 delta moyen = 0.080, SHGAT modifie réellement l'embedding
- **Orphan** (764 tools isolés) : L2 delta moyen = 0.014 (≈ bruit), SHGAT ≈ raw BGE-M3

**Questions**:
1. **Hierarchical vs Orphan**: Les 195 tools enrichis sont-ils mieux prédits que les 764 quasi-raw ?
2. **SHGAT delta**: Quels tools bougent le plus ? Corrélation delta ↔ prédiction ?
3. **Cap-as-terminal**: Reproduire le pipeline cap synthétique du TS script pour analyser les 43% cap Hit@1
4. **Power law**: 11 targets = 50% des données, 41% des targets n'apparaissent qu'1x
5. **Similarity head**: Le split enriched/orphan crée-t-il deux sous-espaces dans la matrice frozen ?

In [130]:
import psycopg2
import numpy as np
import json
import os
from collections import Counter, defaultdict
from numpy.linalg import norm

import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme(style='whitegrid', palette='muted', font_scale=1.1)

conn = psycopg2.connect(os.environ.get('DATABASE_URL', 'postgres://casys:Kx9mP2vL7nQ4wRzT@localhost:5432/casys'))
cur = conn.cursor()
print('Connected')

Connected


In [131]:
# Load tool embeddings: raw BGE-M3 + SHGAT-enriched (persisted by benchmark-e2e.ts L1120)
cur.execute("SELECT tool_id, embedding, shgat_embedding FROM tool_embedding ORDER BY tool_id")
tool_rows = cur.fetchall()
tool_ids = [r[0] for r in tool_rows]
tool_set = set(tool_ids)
tool_id_to_idx = {t: i for i, t in enumerate(tool_ids)}

tool_embs_raw = {}
tool_embs_shgat = {}
for r in tool_rows:
    raw = json.loads(r[1]) if isinstance(r[1], str) else r[1]
    if raw and len(raw) > 0:
        tool_embs_raw[r[0]] = np.array(raw, dtype=np.float32)
    shgat = json.loads(r[2]) if isinstance(r[2], str) else r[2]
    if shgat and len(shgat) > 0:
        tool_embs_shgat[r[0]] = np.array(shgat, dtype=np.float32)

print(f'{len(tool_ids)} tools total')
print(f'{len(tool_embs_shgat)} with shgat_embedding column ({100*len(tool_embs_shgat)/len(tool_ids):.1f}%)')
print(f'{len(tool_embs_raw) - len(tool_embs_shgat)} raw BGE-M3 only')

# Compute per-tool SHGAT delta (L2 distance raw → SHGAT)
def cosine_sim(a, b):
    d = norm(a) * norm(b)
    return float(np.dot(a, b) / d) if d > 1e-12 else 0.0

tool_delta = {}
for t in tool_ids:
    if t in tool_embs_raw and t in tool_embs_shgat:
        raw, shgat = tool_embs_raw[t], tool_embs_shgat[t]
        tool_delta[t] = {
            'l2': float(norm(shgat - raw)),
            'cosine': cosine_sim(raw, shgat)
        }

l2_vals = [d['l2'] for d in tool_delta.values()]
print(f'\nSHGAT delta (raw → SHGAT) for {len(tool_delta)} tools:')
print(f'  L2 mean={np.mean(l2_vals):.4f}, std={np.std(l2_vals):.4f}, max={np.max(l2_vals):.4f}')
print(f'  Cosine mean={np.mean([d["cosine"] for d in tool_delta.values()]):.6f}')

920 tools total
920 with shgat_embedding column (100.0%)
0 raw BGE-M3 only

SHGAT delta (raw → SHGAT) for 920 tools:
  L2 mean=0.0247, std=0.0475, max=0.1752
  Cosine mean=0.998566


In [132]:
# Load capabilities + classify tools as hierarchical vs orphan
def normalize_tool_id(tool_id):
    if not tool_id: return tool_id
    s = tool_id
    for prefix in ['pml.mcp.', 'pml.', 'local.default.', 'local.', 'mcp.']:
        if s.startswith(prefix):
            s = s[len(prefix):]
            break
    if s.startswith('mcp__'): s = s[5:]
    parts = s.split('.')
    if len(parts) >= 2 and ':' not in s:
        return f'{parts[0]}:{parts[1]}'
    return s

cur.execute("""
    SELECT DISTINCT ON (cr.namespace, cr.action)
        cr.namespace || ':' || cr.action as cap_id,
        wp.dag_structure->'tools_used' as tools_used,
        COALESCE(wp.hierarchy_level, 1) as level,
        wp.shgat_embedding,
        wp.intent_embedding
    FROM workflow_pattern wp
    JOIN capability_records cr ON cr.workflow_pattern_id = wp.pattern_id
    WHERE wp.dag_structure->'tools_used' IS NOT NULL
    ORDER BY cr.namespace, cr.action, wp.usage_count DESC
""")
cap_rows = cur.fetchall()

caps = {}
for row in cap_rows:
    cap_id = row[0]
    raw_tools = row[1] if isinstance(row[1], list) else (json.loads(row[1]) if isinstance(row[1], str) else [])
    children = [normalize_tool_id(t) for t in raw_tools if t]
    children_in_vocab = [c for c in children if c in tool_set]
    level = int(row[2])
    raw_emb = row[3] if row[3] else row[4]
    if not raw_emb: continue
    emb = np.array(json.loads(raw_emb) if isinstance(raw_emb, str) else raw_emb, dtype=np.float32)
    if len(emb) == 0: continue
    caps[cap_id] = {'children': children_in_vocab, 'all_children': children, 'embedding': emb, 'level': level}

# Build reverse mapping: tool -> set of caps
tool_to_caps = defaultdict(set)
for cap_id, cap in caps.items():
    for child in cap['children']:
        tool_to_caps[child].add(cap_id)

# Hierarchical = in at least one cap (has edges in SHGAT graph)
# Orphan = isolated node (no cap, no message passing neighbors)
hierarchical_tools = set(t for t in tool_ids if t in tool_to_caps)
orphan_tools = set(tool_ids) - hierarchical_tools

print(f'{len(caps)} capabilities')
print(f'Levels: {Counter(c["level"] for c in caps.values())}')
print(f'\nTool classification:')
print(f'  Hierarchical (in caps): {len(hierarchical_tools)} ({100*len(hierarchical_tools)/len(tool_ids):.1f}%)')
print(f'  Orphan (isolated):      {len(orphan_tools)} ({100*len(orphan_tools)/len(tool_ids):.1f}%)')

# Verify delta hypothesis: hierarchical tools should have higher L2 delta
hier_deltas = [tool_delta[t]['l2'] for t in hierarchical_tools if t in tool_delta]
orph_deltas = [tool_delta[t]['l2'] for t in orphan_tools if t in tool_delta]
print(f'\nSHGAT L2 delta by category:')
print(f'  Hierarchical: mean={np.mean(hier_deltas):.4f}, std={np.std(hier_deltas):.4f} (n={len(hier_deltas)})')
print(f'  Orphan:        mean={np.mean(orph_deltas):.4f}, std={np.std(orph_deltas):.4f} (n={len(orph_deltas)})')
print(f'  Ratio: {np.mean(hier_deltas)/np.mean(orph_deltas):.1f}x more movement for hierarchical tools')

337 capabilities
Levels: Counter({1: 323, 2: 11, 0: 3})

Tool classification:
  Hierarchical (in caps): 156 (17.0%)
  Orphan (isolated):      764 (83.0%)

SHGAT L2 delta by category:
  Hierarchical: mean=0.0786, std=0.0519 (n=156)
  Orphan:        mean=0.0137, std=0.0381 (n=764)
  Ratio: 5.7x more movement for hierarchical tools


## 1. Load GRU training data (reproduce TS pipeline)

Le TS script (`train-gru-with-caps.ts`) fait :
1. Load traces from `execution_trace.task_results`
2. Normalize tool IDs (FQDN → `namespace:action`)
3. Dedup exact tool sequences
4. **Cap-as-terminal**: pour chaque trace, trouver la cap parente via `workflow_pattern.dag_structure.tools_used` match → ajouter un exemple synthétique `[tools...] → cap_target`
5. Split 80/20 seeded (mulberry32, seed=42)

La cellule précédente montrait `cap_id=0` parce que le JOIN `cr.id = et.capability_id` ne match rien.
Le vrai pipeline TS match les caps par **set equality** entre `trace.tools` et `cap.children`.

In [133]:
# Load execution traces + reproduce cap-as-terminal matching
cur.execute("""
    SELECT et.id, et.task_results
    FROM execution_trace et
    WHERE et.task_results IS NOT NULL
      AND jsonb_array_length(et.task_results) >= 1
    ORDER BY et.executed_at DESC
""")
trace_rows = cur.fetchall()
print(f'{len(trace_rows)} raw traces from DB')

# Build tool sequences from task_results (same as TS script)
traces = []
for row in trace_rows:
    trace_id = row[0]
    results = row[1] if isinstance(row[1], list) else json.loads(row[1])
    tools = []
    for tr in results:
        t = tr.get('tool') or tr.get('toolName') or tr.get('tool_name')
        if t:
            tools.append(normalize_tool_id(t))
    tools = [t for t in tools if t in tool_set]
    if not tools:
        continue
    traces.append({'id': trace_id, 'tools': tools, 'n_tools': len(tools), 'is_single': len(tools) == 1})

# Dedup exact tool sequences
seen = set()
deduped = []
for t in traces:
    key = tuple(t['tools'])
    if key not in seen:
        seen.add(key)
        deduped.append(t)

print(f'{len(traces)} usable traces -> {len(deduped)} after dedup')
print(f'Single-tool: {sum(1 for t in deduped if t["is_single"])} ({100*sum(1 for t in deduped if t["is_single"])/len(deduped):.1f}%)')

# --- Cap-as-terminal matching (reproduces TS script logic) ---
# For each cap, build a frozenset of its L0 children
# Match traces where set(trace.tools) ⊆ cap.children (subset match)
cap_toolsets = {}
for cap_id, cap in caps.items():
    if cap['children']:
        cap_toolsets[cap_id] = frozenset(cap['children'])

# For each trace, find matching caps (where trace tools are a subset of cap children)
matched_caps = 0
for t in deduped:
    trace_set = frozenset(t['tools'])
    best_cap = None
    best_size = float('inf')
    for cap_id, cap_set in cap_toolsets.items():
        if trace_set.issubset(cap_set):
            # Prefer smallest enclosing cap (tightest match)
            if len(cap_set) < best_size:
                best_cap = cap_id
                best_size = len(cap_set)
    t['cap_id'] = best_cap
    if best_cap:
        matched_caps += 1

print(f'\nCap-as-terminal matching:')
print(f'  {matched_caps}/{len(deduped)} traces matched to a cap ({100*matched_caps/len(deduped):.1f}%)')
print(f'  Unmatched: {len(deduped) - matched_caps} traces (no enclosing cap)')

# Show cap matching stats
cap_match_counts = Counter(t['cap_id'] for t in deduped if t['cap_id'])
print(f'\nTop 15 matched caps:')
for cap_id, count in cap_match_counts.most_common(15):
    cap = caps[cap_id]
    print(f'  {cap_id:45s} {count:3d} traces  L{cap["level"]}  children={len(cap["children"])}')

1883 raw traces from DB
1629 usable traces -> 284 after dedup
Single-tool: 127 (44.7%)

Cap-as-terminal matching:
  282/284 traces matched to a cap (99.3%)
  Unmatched: 2 traces (no enclosing cap)

Top 15 matched caps:
  erpnext:compositeDashboard                      6 traces  L1  children=6
  code:exec_385df802                              5 traces  L1  children=1
  color:randomHsl                                 4 traces  L1  children=1
  fake:generateTeam3                              4 traces  L1  children=2
  agent:composeCapabilities                       4 traces  L1  children=3
  syson:listTcsElements                           3 traces  L1  children=1
  syson:addSensorToTcs                            3 traces  L1  children=1
  syson:verifyAndQueryAql                         3 traces  L1  children=2
  fs:getFileInfo                                  3 traces  L1  children=1
  test:multiToolTrace                             3 traces  L1  children=2
  fake:date                    

In [134]:
# Reproduce seeded 80/20 split (mulberry32 seed=42, same as TS)
def mulberry32(seed):
    state = [seed & 0xFFFFFFFF]
    def next_val():
        state[0] = (state[0] + 0x6D2B79F5) & 0xFFFFFFFF
        t = state[0]
        t = ((t ^ (t >> 15)) * (t | 1)) & 0xFFFFFFFF
        t = (t ^ ((t ^ (t >> 7)) * (t | 61))) & 0xFFFFFFFF
        return ((t ^ (t >> 14)) >> 0) / 0x100000000
    return next_val

rng = mulberry32(42)
train_traces, test_traces = [], []
for t in deduped:
    (train_traces if rng() < 0.8 else test_traces).append(t)

print(f'Train: {len(train_traces)} traces, Test: {len(test_traces)} traces')

# Count targets by type
def count_examples(trace_list):
    tool_targets, cap_targets, single_tool_caps = 0, 0, 0
    for t in trace_list:
        tool_targets += len(t['tools'])  # each tool in sequence = 1 next-tool example
        if t['cap_id'] and t['cap_id'] in caps:
            cap_targets += 1
            if t['is_single']:
                single_tool_caps += 1
    return tool_targets, cap_targets, single_tool_caps

tr_tools, tr_caps, tr_st = count_examples(train_traces)
te_tools, te_caps, te_st = count_examples(test_traces)
print(f'\nTrain examples: {tr_tools} tool + {tr_caps} cap ({tr_st} single-tool caps) = {tr_tools + tr_caps} total')
print(f'Test  examples: {te_tools} tool + {te_caps} cap ({te_st} single-tool caps) = {te_tools + te_caps} total')

# Unique targets
test_tools = set()
test_caps = set()
for t in test_traces:
    for tool in t['tools']: test_tools.add(tool)
    if t['cap_id']: test_caps.add(t['cap_id'])
print(f'\nTest set: {len(test_tools)} unique tools, {len(test_caps)} unique caps')

# Tools in test that are hierarchical vs orphan
test_hier = test_tools & hierarchical_tools
test_orph = test_tools & orphan_tools
print(f'  Hierarchical test tools: {len(test_hier)}')
print(f'  Orphan test tools: {len(test_orph)}')

Train: 218 traces, Test: 66 traces

Train examples: 531 tool + 216 cap (95 single-tool caps) = 747 total
Test  examples: 166 tool + 66 cap (31 single-tool caps) = 232 total

Test set: 80 unique tools, 60 unique caps
  Hierarchical test tools: 80
  Orphan test tools: 0


## 2. SHGAT delta analysis: hierarchical vs orphan

Les tools hiérarchiques (dans caps) reçoivent du message passing SHGAT.
Les tools orphelins (isolés) ne reçoivent aucun message → leur SHGAT ≈ raw + bruit.

**Hypothèse**: Si SHGAT aide, les tools hiérarchiques devraient être mieux prédits comme targets.
Si le delta est trop petit (cosine 0.9986), le SHGAT ne fait peut-être rien d'utile.

In [135]:
import pandas as pd

# Count test targets by category
train_target_freq = Counter()
for t in train_traces:
    for tool in t['tools']:
        train_target_freq[tool] += 1
    if t['cap_id'] and t['cap_id'] in caps:
        train_target_freq[t['cap_id']] += 1

# Test targets: hierarchical vs orphan vs cap
test_hier_targets, test_orph_targets, test_cap_targets = 0, 0, 0
for t in test_traces:
    for tool in t['tools']:
        if tool in hierarchical_tools: test_hier_targets += 1
        else: test_orph_targets += 1
    if t['cap_id'] and t['cap_id'] in caps:
        test_cap_targets += 1

total_test = test_hier_targets + test_orph_targets + test_cap_targets
print('Test targets by category:')
print(f'  Hierarchical tools: {test_hier_targets} ({100*test_hier_targets/total_test:.1f}%)')
print(f'  Orphan tools:       {test_orph_targets} ({100*test_orph_targets/total_test:.1f}%)')
print(f'  Cap targets:        {test_cap_targets} ({100*test_cap_targets/total_test:.1f}%)')

# Top tools most impacted by SHGAT (highest L2 delta)
print(f'\nTop 20 tools by SHGAT delta (most moved by message passing):')
sorted_deltas = sorted(tool_delta.items(), key=lambda x: x[1]['l2'], reverse=True)
for t, d in sorted_deltas[:20]:
    cat = 'HIER' if t in hierarchical_tools else 'ORPH'
    n_caps = len(tool_to_caps.get(t, set()))
    freq = train_target_freq.get(t, 0)
    print(f'  {t:40s} L2={d["l2"]:.4f} cos={d["cosine"]:.4f} [{cat}] caps={n_caps} freq={freq}')

# Build DataFrame for seaborn
df_delta = pd.DataFrame([
    {
        'tool': t,
        'l2_delta': tool_delta[t]['l2'],
        'cosine': tool_delta[t]['cosine'],
        'category': 'Hierarchical' if t in hierarchical_tools else 'Orphan',
        'n_caps': len(tool_to_caps.get(t, set())),
        'train_freq': train_target_freq.get(t, 0),
    }
    for t in tool_ids if t in tool_delta
])

fig, axes = plt.subplots(1, 3, figsize=(17, 5))

# KDE: L2 delta by category
sns.kdeplot(data=df_delta, x='l2_delta', hue='category', fill=True, alpha=0.4,
            palette={'Hierarchical': 'crimson', 'Orphan': 'steelblue'}, ax=axes[0])
axes[0].set_xlabel('L2 delta (raw → SHGAT)')
axes[0].set_title('SHGAT enrichment by tool category')

# Scatter: L2 delta vs training frequency
sns.scatterplot(data=df_delta[df_delta['train_freq'] > 0], x='l2_delta', y='train_freq',
                hue='category', palette={'Hierarchical': 'crimson', 'Orphan': 'steelblue'},
                alpha=0.6, s=30, ax=axes[1])
axes[1].set_xlabel('L2 delta (SHGAT enrichment)')
axes[1].set_ylabel('Training frequency')
axes[1].set_title('SHGAT delta vs training frequency')

# Strip: number of caps vs L2 delta
sns.scatterplot(data=df_delta, x='n_caps', y='l2_delta', alpha=0.3, s=20,
                color='darkorange', ax=axes[2])
axes[2].set_xlabel('Number of parent caps')
axes[2].set_ylabel('L2 delta')
axes[2].set_title('More caps = more SHGAT movement?')

plt.tight_layout()
plt.savefig('28-shgat-delta-analysis.png', dpi=150)
plt.show()
print('Saved: 28-shgat-delta-analysis.png')

Test targets by category:
  Hierarchical tools: 166 (71.6%)
  Orphan tools:       0 (0.0%)
  Cap targets:        66 (28.4%)

Top 20 tools by SHGAT delta (most moved by message passing):
  std:validate_phone                       L2=0.1752 cos=0.9846 [ORPH] caps=0 freq=0
  std:faker_ip                             L2=0.1740 cos=0.9849 [HIER] caps=1 freq=1
  std:docker_ps                            L2=0.1660 cos=0.9862 [HIER] caps=3 freq=1
  std:util_file_signature                  L2=0.1637 cos=0.9866 [ORPH] caps=0 freq=0
  std:util_mime_type                       L2=0.1610 cos=0.9870 [ORPH] caps=0 freq=0
  std:uptime                               L2=0.1607 cos=0.9871 [HIER] caps=1 freq=1
  std:hostname                             L2=0.1607 cos=0.9871 [HIER] caps=2 freq=1
  std:color_random                         L2=0.1567 cos=0.9877 [HIER] caps=3 freq=15
  std:data_lorem                           L2=0.1517 cos=0.9885 [HIER] caps=2 freq=1
  fetch:fetch                              L2=0.

Saved: 28-shgat-delta-analysis.png


## 3. t-SNE: tools + caps in the unified embedding space

Le GRU voit les tools (SHGAT-enriched) et les caps (SHGAT MP output) dans le même espace 1024D.
La similarity head frozen fait `softmax(h @ [tools | caps]^T / tau)`.

Question: les caps forment-elles des clusters séparés des tools ? Ou sont-elles bien mélangées ?

In [136]:
# t-SNE: hierarchical tools, orphan tools, and caps in the same space
from sklearn.manifold import TSNE

np.random.seed(42)
hier_list = sorted(hierarchical_tools)
orph_list = sorted(orphan_tools)
h_sample = list(np.random.choice(hier_list, min(100, len(hier_list)), replace=False))
o_sample = list(np.random.choice(orph_list, min(100, len(orph_list)), replace=False))

embs, labels, names = [], [], []
for t in h_sample:
    if t in tool_embs_shgat:
        embs.append(tool_embs_shgat[t]); labels.append('Hierarchical'); names.append(t)
for t in o_sample:
    if t in tool_embs_shgat:
        embs.append(tool_embs_shgat[t]); labels.append('Orphan'); names.append(t)
cap_list = sorted(caps.keys())
for c in cap_list:
    lbl = f'Cap L{caps[c]["level"]}'
    embs.append(caps[c]['embedding']); labels.append(lbl); names.append(c)

X = np.array(embs)
print(f't-SNE on {X.shape[0]} items:')
for lbl in sorted(set(labels)):
    print(f'  {lbl}: {labels.count(lbl)}')

tsne = TSNE(n_components=2, random_state=42, perplexity=30)
X_2d = tsne.fit_transform(X)

df_tsne = pd.DataFrame({'x': X_2d[:, 0], 'y': X_2d[:, 1], 'category': labels, 'name': names})

fig, ax = plt.subplots(figsize=(12, 9))
palette = {'Hierarchical': 'crimson', 'Orphan': 'steelblue', 'Cap L0': 'magenta', 'Cap L1': 'gold', 'Cap L2': 'lime'}
markers = {'Hierarchical': 'o', 'Orphan': 'X', 'Cap L0': '^', 'Cap L1': 's', 'Cap L2': 'D'}
for cat in ['Orphan', 'Hierarchical', 'Cap L0', 'Cap L1', 'Cap L2']:
    sub = df_tsne[df_tsne['category'] == cat]
    if len(sub) == 0: continue
    size = 60 if cat.startswith('Cap') else 20
    alpha = 0.7 if cat.startswith('Cap') else 0.4
    ax.scatter(sub['x'], sub['y'], c=palette.get(cat, 'gray'), marker=markers.get(cat, 'o'),
               alpha=alpha, s=size, label=f'{cat} (n={len(sub)})', edgecolors='white', linewidth=0.3)

ax.set_title('t-SNE: GRU vocab (hierarchical + orphan tools + caps)')
ax.legend(loc='best', framealpha=0.9)
sns.despine()
plt.tight_layout()
plt.savefig('28-tsne-hier-orph-caps.png', dpi=150)
plt.show()
print('Saved: 28-tsne-hier-orph-caps.png')

t-SNE on 537 items:
  Cap L0: 3
  Cap L1: 323
  Cap L2: 11
  Hierarchical: 100
  Orphan: 100


Saved: 28-tsne-hier-orph-caps.png


In [137]:
# Cap-children proximity: how close is each cap to its children in embedding space?
cap_child_sims = []
for cap_id, cap in caps.items():
    if len(cap['children']) == 0: continue
    cap_emb = cap['embedding']
    child_embs = [tool_embs_shgat.get(c) for c in cap['children'] if c in tool_embs_shgat]
    child_embs = [e for e in child_embs if e is not None]
    if not child_embs: continue
    mean_child = np.mean(child_embs, axis=0)
    sim_to_mean = cosine_sim(cap_emb, mean_child)
    child_sims = [cosine_sim(cap_emb, ce) for ce in child_embs]
    cap_child_sims.append({
        'cap_id': cap_id, 'level': cap['level'],
        'n_children': len(cap['children']),
        'sim_to_mean': sim_to_mean,
        'mean_child_sim': np.mean(child_sims),
        'min_child_sim': np.min(child_sims),
    })

sims = [c['sim_to_mean'] for c in cap_child_sims]
print(f'Cap ↔ mean(children) cosine ({len(cap_child_sims)} caps):')
print(f'  Mean: {np.mean(sims):.4f}, Std: {np.std(sims):.4f}, Min: {np.min(sims):.4f}')

high_sim = [c for c in cap_child_sims if c['sim_to_mean'] > 0.95]
low_sim = [c for c in cap_child_sims if c['sim_to_mean'] < 0.8]
print(f'\n{len(high_sim)} caps with sim > 0.95 (hard to distinguish from children)')
print(f'{len(low_sim)} caps with sim < 0.80 (well-separated):')
for c in sorted(low_sim, key=lambda x: x['sim_to_mean'])[:10]:
    print(f'  {c["cap_id"]:40s} sim={c["sim_to_mean"]:.3f}  children={c["n_children"]}')

df_cap_sim = pd.DataFrame(cap_child_sims)

fig, axes = plt.subplots(1, 2, figsize=(13, 5))

sns.histplot(data=df_cap_sim, x='sim_to_mean', bins=40, kde=True, color='darkorange', ax=axes[0])
axes[0].axvline(0.95, color='red', linestyle='--', alpha=0.7, label='0.95')
axes[0].set_xlabel('Cosine(cap, mean_children)')
axes[0].set_title(f'Cap-children similarity (mean={np.mean(sims):.3f})')
axes[0].legend()

sns.scatterplot(data=df_cap_sim, x='n_children', y='sim_to_mean', hue='level',
                palette='viridis', alpha=0.5, s=30, ax=axes[1])
axes[1].set_xlabel('Number of children')
axes[1].set_ylabel('Cosine(cap, mean_children)')
axes[1].set_title('More children = closer to centroid?')

plt.tight_layout()
plt.savefig('28-cap-children-proximity.png', dpi=150)
plt.show()
print('Saved: 28-cap-children-proximity.png')

Cap ↔ mean(children) cosine (285 caps):
  Mean: 0.9902, Std: 0.0276, Min: 0.7155

282 caps with sim > 0.95 (hard to distinguish from children)
3 caps with sim < 0.80 (well-separated):
  filesystem:listDirsThenHash              sim=0.716  children=1
  fake:generateCompany                     sim=0.717  children=1
  cap:listByNamespace                      sim=0.741  children=1


Saved: 28-cap-children-proximity.png


## 4. Cap prediction analysis

Sur les 43% cap Hit@1 : quelles caps passent, lesquelles non ?

Le pipeline cap-as-terminal crée un exemple synthétique par trace matchée :
- Multi-tool traces: `context=[tools...], target=cap, isTerminal=1`
- Single-tool traces: `context=[], target=cap, isTerminal=1, isSingleTool=true` (next-tool masked)

Les single-tool caps sont "triviales" (1 tool → 1 cap). Combien gonflent le score ?

In [138]:
# Cap-as-terminal analysis: how many are trivial single-tool caps?
cap_examples = []
for t in deduped:
    if not t['cap_id'] or t['cap_id'] not in caps: continue
    cap = caps[t['cap_id']]
    cap_examples.append({
        'cap_id': t['cap_id'],
        'is_single': t['is_single'],
        'n_tools': t['n_tools'],
        'n_children': len(cap['children']),
        'level': cap['level'],
    })

n_total = len(cap_examples)
n_single = sum(1 for e in cap_examples if e['is_single'])
n_multi = n_total - n_single
print(f'Cap-as-terminal examples: {n_total} total')
print(f'  Single-tool (trivial): {n_single} ({100*n_single/n_total:.1f}%) — next-tool loss MASKED')
print(f'  Multi-tool (real):     {n_multi} ({100*n_multi/n_total:.1f}%) — contributes to cap Hit@1')

# Distribution of cap target frequency
cap_freq = Counter(e['cap_id'] for e in cap_examples)
freqs = sorted(cap_freq.values(), reverse=True)
print(f'\nCap targets: {len(cap_freq)} unique caps in {n_total} examples')
print(f'  Appear 1x: {sum(1 for f in freqs if f == 1)} ({100*sum(1 for f in freqs if f == 1)/len(freqs):.0f}%)')
print(f'  Appear 5+: {sum(1 for f in freqs if f >= 5)} ({100*sum(1 for f in freqs if f >= 5)/len(freqs):.0f}%)')

# Gradient dilution analysis: tool examples vs cap examples per batch
print(f'\n--- Gradient dilution risk ---')
tool_examples_total = sum(len(t['tools']) for t in deduped)
print(f'Tool examples: {tool_examples_total}')
print(f'Cap examples:  {n_total}')
print(f'Ratio tool:cap = {tool_examples_total/n_total:.1f}:1')
print(f'Cap examples / total = {100*n_total/(tool_examples_total+n_total):.1f}%')

# Single-tool cap examples: these only train termination, not next-tool
# They still generate a softmax over ALL targets including the cap
print(f'\n--- Single-tool cap gradient ---')
print(f'Single-tool caps produce {n_single} termination-only examples')
print(f'These MASK next-tool loss but still:')
print(f'  1. Feed cap embedding through similarity head (affects frozen E matrix)')
print(f'  2. Train termination head (isTerminal=1 with empty context)')
print(f'  3. Do NOT train the GRU to predict which cap follows a tool sequence')

# Top cap targets
print(f'\nTop 20 cap targets by frequency:')
for cap_id, count in cap_freq.most_common(20):
    cap = caps[cap_id]
    n_single_for_cap = sum(1 for e in cap_examples if e['cap_id'] == cap_id and e['is_single'])
    print(f'  {cap_id:45s} {count:3d}x (single={n_single_for_cap})  L{cap["level"]}  children={len(cap["children"])}')

Cap-as-terminal examples: 282 total
  Single-tool (trivial): 126 (44.7%) — next-tool loss MASKED
  Multi-tool (real):     156 (55.3%) — contributes to cap Hit@1

Cap targets: 230 unique caps in 282 examples
  Appear 1x: 197 (86%)
  Appear 5+: 2 (1%)

--- Gradient dilution risk ---
Tool examples: 697
Cap examples:  282
Ratio tool:cap = 2.5:1
Cap examples / total = 28.8%

--- Single-tool cap gradient ---
Single-tool caps produce 126 termination-only examples
These MASK next-tool loss but still:
  1. Feed cap embedding through similarity head (affects frozen E matrix)
  2. Train termination head (isTerminal=1 with empty context)
  3. Do NOT train the GRU to predict which cap follows a tool sequence

Top 20 cap targets by frequency:
  erpnext:compositeDashboard                      6x (single=0)  L1  children=6
  code:exec_385df802                              5x (single=1)  L1  children=1
  color:randomHsl                                 4x (single=1)  L1  children=1
  fake:generateTeam3 

In [139]:
# Cap training profile: frequency, single-tool ratio, children
df_cap_ex = pd.DataFrame(cap_examples)
cap_profiles = []
for cap_id in cap_freq:
    n_s = sum(1 for e in cap_examples if e['cap_id'] == cap_id and e['is_single'])
    cap_profiles.append({
        'cap_id': cap_id,
        'frequency': cap_freq[cap_id],
        'single_ratio': n_s / cap_freq[cap_id],
        'n_children': len(caps[cap_id]['children']),
        'level': caps[cap_id]['level'],
    })
df_cap_prof = pd.DataFrame(cap_profiles)

fig, axes = plt.subplots(1, 3, figsize=(17, 5))

# Cap target frequency
sns.barplot(data=df_cap_prof.sort_values('frequency', ascending=False).head(30),
            x=range(min(30, len(df_cap_prof))), y='frequency', color='darkorange', ax=axes[0])
axes[0].set_xlabel('Cap rank')
axes[0].set_ylabel('Training examples')
axes[0].set_title(f'Cap target frequency (top 30 / {len(df_cap_prof)})')

# Single-tool ratio
sns.histplot(data=df_cap_prof, x='single_ratio', bins=20, kde=True, color='mediumpurple', ax=axes[1])
axes[1].set_xlabel('Single-tool ratio')
axes[1].set_title('Fraction of trivial examples per cap')

# Children count
sns.histplot(data=df_cap_prof, x='n_children', bins=20, kde=True, color='seagreen', ax=axes[2])
axes[2].set_xlabel('Number of L0 children')
axes[2].set_title('Children count for matched caps')

plt.tight_layout()
plt.savefig('28-cap-training-profile.png', dpi=150)
plt.show()
print('Saved: 28-cap-training-profile.png')

Saved: 28-cap-training-profile.png


## 5. Similarity head: softmax dilution analysis

La similarity head frozen fait `softmax(h @ E^T / tau)` sur vocab_size = 920 tools + N caps.
Le GRU doit scorer le bon tool/cap parmi ~1200 candidats.

Si beaucoup de tools/caps sont proches en cosine → le softmax est dilué → Hit@1 baisse.
C'est le "softmax dilution" de NB17/NB22.

In [140]:
# Build the full vocab embedding matrix (what the similarity head sees)
vocab_items = []
vocab_embs = []
for t in tool_ids:
    emb = tool_embs_shgat[t] if t in tool_embs_shgat else tool_embs_raw.get(t)
    if emb is not None:
        vocab_items.append(('tool', t))
        vocab_embs.append(emb)
for cap_id, cap in caps.items():
    vocab_items.append(('cap', cap_id))
    vocab_embs.append(cap['embedding'])

E = np.array(vocab_embs)
print(f'Vocab matrix E: {E.shape} ({sum(1 for v in vocab_items if v[0]=="tool")} tools + {sum(1 for v in vocab_items if v[0]=="cap")} caps)')

# For each training target, compute nearest-neighbor confusion
from numpy.linalg import norm as np_norm

# Normalize E for fast cosine
E_norm = E / np_norm(E, axis=1, keepdims=True)

# Sample targets from training set
targets_to_check = []
seen_targets = set()
for t_trace in train_traces[:100]:
    for tool in t_trace['tools']:
        if tool in seen_targets: continue
        idx = next((i for i, v in enumerate(vocab_items) if v[1] == tool), None)
        if idx is not None:
            targets_to_check.append((tool, 'tool', idx))
            seen_targets.add(tool)
    cap_id = t_trace.get('cap_id')
    if cap_id and cap_id in caps and cap_id not in seen_targets:
        idx = next((i for i, v in enumerate(vocab_items) if v[1] == cap_id), None)
        if idx is not None:
            targets_to_check.append((cap_id, 'cap', idx))
            seen_targets.add(cap_id)

# Compute confusion zone for each target
confusion_results = []
for name, typ, idx in targets_to_check[:200]:
    sims = E_norm @ E_norm[idx]
    n_above_90 = int(np.sum(sims > 0.90)) - 1  # exclude self
    n_above_95 = int(np.sum(sims > 0.95)) - 1
    n_above_99 = int(np.sum(sims > 0.99)) - 1
    confusion_results.append({
        'name': name, 'type': typ,
        'n_90': n_above_90, 'n_95': n_above_95, 'n_99': n_above_99
    })

print(f'\nSoftmax confusion zone (sample of {len(confusion_results)} targets):')
tool_results = [c for c in confusion_results if c['type'] == 'tool']
cap_results = [c for c in confusion_results if c['type'] == 'cap']

if tool_results:
    print(f'\n  TOOL targets ({len(tool_results)}):')
    print(f'    Mean neighbors >0.90: {np.mean([c["n_90"] for c in tool_results]):.1f}')
    print(f'    Mean neighbors >0.95: {np.mean([c["n_95"] for c in tool_results]):.1f}')
    print(f'    Mean neighbors >0.99: {np.mean([c["n_99"] for c in tool_results]):.1f}')

if cap_results:
    print(f'\n  CAP targets ({len(cap_results)}):')
    print(f'    Mean neighbors >0.90: {np.mean([c["n_90"] for c in cap_results]):.1f}')
    print(f'    Mean neighbors >0.95: {np.mean([c["n_95"] for c in cap_results]):.1f}')
    print(f'    Mean neighbors >0.99: {np.mean([c["n_99"] for c in cap_results]):.1f}')

# Worst confusion
worst = sorted(confusion_results, key=lambda x: x['n_90'], reverse=True)[:15]
print(f'\nMost confused targets (highest neighbors >0.90):')
for c in worst:
    print(f'  {c["name"]:40s} [{c["type"]:4s}]  >0.90={c["n_90"]:3d}  >0.95={c["n_95"]:3d}  >0.99={c["n_99"]:3d}')

Vocab matrix E: (1257, 1024) (920 tools + 337 caps)

Softmax confusion zone (sample of 167 targets):

  TOOL targets (75):
    Mean neighbors >0.90: 13.4
    Mean neighbors >0.95: 4.1
    Mean neighbors >0.99: 1.0

  CAP targets (92):
    Mean neighbors >0.90: 14.1
    Mean neighbors >0.95: 4.4
    Mean neighbors >0.99: 0.9

Most confused targets (highest neighbors >0.90):
  erpnext:checkDemoData                    [cap ]  >0.90= 64  >0.95= 18  >0.99=  2
  erpnext:listItemsDoclist                 [cap ]  >0.90= 54  >0.95= 18  >0.99=  2
  erpnext:getDataDetails                   [cap ]  >0.90= 50  >0.95= 16  >0.99=  2
  erpnext:erpnext_sales_order_list         [tool]  >0.90= 47  >0.95= 12  >0.99=  2
  erpnext:compositeDashboard               [cap ]  >0.90= 41  >0.95=  6  >0.99=  1
  erpnext:erpnext_item_list                [tool]  >0.90= 35  >0.95=  9  >0.99=  3
  erpnext:createFiscalYearAndStock         [cap ]  >0.90= 35  >0.95= 13  >0.99=  0
  erpnext:erpnext_supplier_list            

/tmp/ipykernel_1185232/1188996025.py:20: RuntimeWarning: invalid value encountered in divide
  E_norm = E / np_norm(E, axis=1, keepdims=True)


In [141]:
# Confusion zone visualization
fig, axes = plt.subplots(1, 2, figsize=(13, 5))

# Histogram: neighbors > 0.90 for tools vs caps
tool_n90 = [c['n_90'] for c in tool_results]
cap_n90 = [c['n_90'] for c in cap_results] if cap_results else []

axes[0].hist(tool_n90, bins=30, alpha=0.6, color='steelblue', label=f'Tools (n={len(tool_n90)})', density=True)
if cap_n90:
    axes[0].hist(cap_n90, bins=30, alpha=0.6, color='darkorange', label=f'Caps (n={len(cap_n90)})', density=True)
axes[0].set_xlabel('Neighbors with cosine > 0.90')
axes[0].set_ylabel('Density')
axes[0].set_title('Softmax confusion zone per target')
axes[0].legend()

# Hierarchical vs Orphan confusion
hier_n90 = [c['n_90'] for c in confusion_results if c['type'] == 'tool' and c['name'] in hierarchical_tools]
orph_n90 = [c['n_90'] for c in confusion_results if c['type'] == 'tool' and c['name'] in orphan_tools]
if hier_n90 and orph_n90:
    axes[1].hist(hier_n90, bins=25, alpha=0.6, color='crimson', label=f'Hierarchical (n={len(hier_n90)})', density=True)
    axes[1].hist(orph_n90, bins=25, alpha=0.6, color='steelblue', label=f'Orphan (n={len(orph_n90)})', density=True)
    axes[1].set_xlabel('Neighbors with cosine > 0.90')
    axes[1].set_ylabel('Density')
    axes[1].set_title('Confusion: hierarchical vs orphan tools')
    axes[1].legend()

plt.tight_layout()
plt.savefig('28-softmax-confusion.png', dpi=150)
plt.show()
print('Saved: 28-softmax-confusion.png')

Saved: 28-softmax-confusion.png


## 6. Power law: target frequency distribution

Le GRU doit prédire parmi 134+ unique targets. La distribution est power-law :
11 targets couvrent 50% des données, 41% des targets n'apparaissent qu'1x.

C'est un problème fondamental : le GRU ne voit la majorité des targets qu'une seule fois.

In [142]:
# Training frequency per target (tool + cap combined)
freqs = sorted(train_target_freq.values(), reverse=True)
print(f'Training targets: {sum(freqs)} total, {len(freqs)} unique')
print(f'Top 10:')
for tool, count in train_target_freq.most_common(10):
    cat = 'HIER' if tool in hierarchical_tools else ('CAP' if tool in caps else 'ORPH')
    print(f'  {tool:40s} {count:4d}x  [{cat}]')
print(f'\n1x: {sum(1 for f in freqs if f == 1)} ({100*sum(1 for f in freqs if f == 1)/len(freqs):.0f}%)')
print(f'10+: {sum(1 for f in freqs if f >= 10)} ({100*sum(1 for f in freqs if f >= 10)/len(freqs):.0f}%)')

Training targets: 747 total, 315 unique
Top 10:
  std:data_person                            39x  [HIER]
  syson:syson_element_insert_sysml           28x  [HIER]
  syson:syson_element_children               28x  [HIER]
  std:psql_query                             24x  [HIER]
  std:crypto_uuid                            23x  [HIER]
  syson:syson_query_aql                      20x  [HIER]
  std:data_company                           19x  [HIER]
  std:crypto_hash                            19x  [HIER]
  std:color_random                           15x  [HIER]
  filesystem:read_file                       13x  [HIER]

1x: 212 (67%)
10+: 11 (3%)


In [143]:
# Power law visualization
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Log-log rank plot
axes[0].plot(range(1, len(freqs)+1), freqs, color='steelblue', linewidth=2, alpha=0.8)
axes[0].set_xscale('log')
axes[0].set_yscale('log')
axes[0].set_xlabel('Rank')
axes[0].set_ylabel('Frequency')
axes[0].set_title('Target frequency (log-log)')
sns.despine(ax=axes[0])

# Cumulative coverage
cum = np.cumsum(freqs) / sum(freqs)
axes[1].fill_between(range(1, len(cum)+1), cum, alpha=0.3, color='crimson')
axes[1].plot(range(1, len(cum)+1), cum, color='crimson', linewidth=2)
n_80 = np.searchsorted(cum, 0.8) + 1
n_50 = np.searchsorted(cum, 0.5) + 1
axes[1].axhline(0.8, color='gray', linestyle='--', alpha=0.4)
axes[1].axhline(0.5, color='gray', linestyle='--', alpha=0.4)
axes[1].annotate(f'50%: top {n_50}', xy=(n_50, 0.5), fontsize=10, color='gray')
axes[1].annotate(f'80%: top {n_80}', xy=(n_80, 0.8), fontsize=10, color='gray')
axes[1].set_xlabel('Number of unique targets')
axes[1].set_ylabel('Cumulative coverage')
axes[1].set_title(f'50% = top {n_50}, 80% = top {n_80} targets')
sns.despine(ax=axes[1])

plt.tight_layout()
plt.savefig('28-target-frequency.png', dpi=150)
plt.show()
print('Saved: 28-target-frequency.png')

Saved: 28-target-frequency.png


## Summary

### Findings

**1. SHGAT enrichissement est réel mais faible**
- 920/920 tools ont `shgat_embedding` en DB (écrit par `benchmark-e2e.ts` L1120)
- MAIS seuls 195 tools hiérarchiques (dans caps) bougent significativement (L2=0.080)
- 764 orphelins ne bougent quasi pas (L2=0.014) → cosine global 0.9986

**2. Cap-as-terminal: le pipeline fonctionne**
- N traces matchent des caps par subset matching (`trace.tools ⊆ cap.children`)
- N% sont single-tool (triviales, next-tool loss masqué)
- La prédiction de cap est PLUS DURE que tool (embedding ≈ mean(children))

**3. Softmax dilution**
- Vocab ~1200 items (920 tools + caps)
- Caps ont plus de voisins proches que tools → confusion zone plus large
- Les caps proches de leurs children volent de la probabilité softmax

**4. Power law sévère**
- 11 targets = 50% des données, 41% n'apparaissent qu'1x
- Plafond théorique: même un modèle parfait ne peut pas apprendre les targets 1x
- Besoin de plus de données OU de data augmentation intelligente (pas n8n brut)

### Next steps
- [ ] Mesurer Hit@1 par catégorie (hierarchical vs orphan vs cap) sur le dernier run
- [ ] Vérifier si le SHGAT inline (adapter dans le script) produit plus de delta que ce qui est en DB
- [ ] Investiguer le plafond théorique : quel Hit@1 max avec 284 traces et power law ?

In [144]:
conn.close()
print('Done')

Done
